### PR TITLE
Sort and group order not updated when using setAttribute.

### PR DESCRIPTION
**Pull request by nh13**
@yfarjoun could you take a look since this fixes a bug you introduced, and @tfenne can a maintainer take a look?

The issue was that folks can call `setAttribute("SO", "queryname")` and the cached sort order introduced in ad8aa02aa2ed02b9cb2bb8801c3e6897e91ec3f6 would not get updated.  Similarly for the group order.

### CODE REVIEW

File: src/main/java/htsjdk/samtools/SAMFileHeader.java [hunk 1]
```diff
@@ -292,7 +292,43 @@ public GroupOrder getGroupOrder() {
...
+     * @deprecated Use {@link #setAttribute(String, String) instead
+     */
+    @Deprecated
+    @Override
+    public void setAttribute(final String key, final Object value) {
+        super.setAttribute(key, value);
+        if (key.equals(SORT_ORDER_TAG) || key.equals(GROUP_ORDER_TAG)) {
+            this.setAttribute(key, value.toString());
+        } else {
+            super.setAttribute(key, value);
```
Comment (pshapiro4broad) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: This is calling `super.setAttribute()` twice is that intentional?

File: src/main/java/htsjdk/samtools/SAMFileHeader.java [hunk 2]
```diff
@@ -292,7 +292,43 @@ public GroupOrder getGroupOrder() {
...
+     * Set the given value for the attribute named 'key'.  Replaces an existing value, if any.
+     * If value is null, the attribute is removed.
+     * Supported types are Character, Integer, Float and String.  Byte and Short may also be
+     * passed in but they will be converted to Integer.
+     * @param key attribute name
+     * @param value attribute value
+     */
+    @Override
+    public void setAttribute(final String key, final String value) {
+        super.setAttribute(key, value);
```
Comment (pshapiro4broad) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: Also called twice here.

File: src/main/java/htsjdk/samtools/SAMFileHeader.java [hunk 3]
```diff
@@ -292,7 +292,43 @@ public GroupOrder getGroupOrder() {
...
+        if (key.equals(SORT_ORDER_TAG)) this.setSortOrder(SortOrder.valueOf(value));
```
Comment (pshapiro4broad) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: This code should use `{` `}` around `if`/`else` as well.

File: src/main/java/htsjdk/samtools/SAMFileHeader.java (hunk 2 above)
Comment (nh13) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: Ugh, left over from trying to make the first commit's test fail.  Good catch.

**Comment by nh13** [comment]
@pshapiro4broad thanks for spotting those!

File: src/main/java/htsjdk/samtools/SAMFileHeader.java [hunk 4]
```diff
@@ -325,10 +324,11 @@ public void setAttribute(final String key, final Object value) {
...
     public void setAttribute(final String key, final String value) {
-        super.setAttribute(key, value);
-        if (key.equals(SORT_ORDER_TAG)) this.setSortOrder(SortOrder.valueOf(value));
-        else if (key.equals(GROUP_ORDER_TAG)) this.setGroupOrder(GroupOrder.valueOf(value));
-        else super.setAttribute(key, value);
+        if (key.equals(SORT_ORDER_TAG)) {
+            this.setSortOrder(SortOrder.valueOf(value));
+        } else if (key.equals(GROUP_ORDER_TAG)) {
+            this.setGroupOrder(GroupOrder.valueOf(value));
+        } else super.setAttribute(key, value);
```
Comment (nh13) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: will @pshapiro4broad allow this, or should I put more brackets?

Comment (nh13) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: I put more brackets

Comment (pshapiro4broad) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: lol thanks @nh13 
I think it's 👍 but someone who actually understands the code should probably look at it too.

**Comment by nh13** [comment]
@yfarjoun this is now breaking other tests (namely `tagCorrectlyProcessTest` in `ValidateSamFileTest`).  I feel like reverting your commit to remove the internal references to sort and group order.  What do you think?

**Comment by nh13** [comment]
@yfarjoun or @tfenne care to take a quick final peak?

File: src/main/java/htsjdk/samtools/SAMFileHeader.java [hunk 5]
```diff
@@ -292,7 +292,44 @@ public GroupOrder getGroupOrder() {
...
+            this.setAttribute(key, value.toString());
+        } else {
+            super.setAttribute(key, value);
+        }
+    }
+
+    /**
+     * Set the given value for the attribute named 'key'.  Replaces an existing value, if any.
+     * If value is null, the attribute is removed.
+     * Supported types are Character, Integer, Float and String.  Byte and Short may also be
```
Comment (tfenne) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: The parameter type is String though?

File: src/test/java/htsjdk/samtools/SAMFileHeaderTest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ */
+package htsjdk.samtools;
+
+import htsjdk.HtsjdkTest;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class SAMFileHeaderTest extends HtsjdkTest {
+
```
Comment (tfenne) [inline src/test/java/htsjdk/samtools/SAMFileHeaderTest.java]: Could you also just add lines to the test that ensure that header.getAttribute("SO"/"GO") do the right things after each of the setting methods?

File: src/main/java/htsjdk/samtools/SAMFileHeader.java (hunk 5 above)
Comment (nh13) [inline src/main/java/htsjdk/samtools/SAMFileHeader.java]: Will fix this here and in `AbstractSAMHeaderRecord` where I copied it from.

File: src/test/java/htsjdk/samtools/SAMFileHeaderTest.java (hunk 6 above)
Comment (nh13) [inline src/test/java/htsjdk/samtools/SAMFileHeaderTest.java]: Of course!

**Comment by nh13** [comment]
@tfenne I think this is ready for a squash and merge after your thumbs up.

**Pull request merged**